rali/rali/source/tf_record_reader.cpp

```diff
@@ -0,0 +1,329 @@
+        _folder_path = subfolder_path;
+        if (tf_record_reader() != Reader::Status::OK)
+            WRN("FileReader ShardID [" + TOSTR(_shard_id) + "] File reader cannot access the storage at " + _folder_path);
+    }
+    if (_in_batch_read_count > 0 && _in_batch_read_count < _batch_count)
+    {
+        replicate_last_image_to_fill_last_shard();
+        LOG("FileReader ShardID [" + TOSTR(_shard_id) + "] Replicated " + _folder_path + _last_file_name + " " + TOSTR((_batch_count - _in_batch_read_count)) + " times to fill the last batch")
+    }
+    if (!_file_names.empty())
+        LOG("FileReader ShardID [" + TOSTR(_shard_id) + "] Total of " + TOSTR(_file_names.size()) + " images loaded from " + _full_path)
+    closedir(_sub_dir);
+    return ret;
+}
+void TFRecordReader::replicate_last_image_to_fill_last_shard()
+{
+    // std::cerr<<"\n Replicate last image";
+    for (size_t i = _in_batch_read_count; i < _batch_count; i++)
+    {
+        _file_names.push_back(_last_file_name);
+        _file_size.insert(std::pair<std::string, unsigned int>(_last_file_name, _last_file_size));
+    }
+}
+
+Reader::Status TFRecordReader::tf_record_reader()
+{
+    std::string fname = _folder_path;
+
+    uint file_size;
+    std::ifstream file_contents(fname.c_str(), std::ios::binary);
+    if(!file_contents)
+        THROW("TFRecordReader: Failed to open file "+fname);
+    file_contents.seekg(0, std::ifstream::end);
+    file_size = file_contents.tellg();
+    // std::cerr<<"\n length of the file:: "<<length<<std::endl;
+    file_contents.seekg(0, std::ifstream::beg);
+    auto ret = read_image_names(file_contents, file_size);
+    if(ret != Reader::Status::OK )
+        THROW("TFRecordReader: Error in reading TF records");
+    _last_rec = false;
+    if (_file_names.size() != _file_size.size())
+        std::cerr << "\n Size of vectors are not same";
+    file_contents.close();
+
+    return Reader::Status::OK;
+}
+
+size_t TFRecordReader::get_file_shard_id()
+{
+    if (_batch_count == 0 || _shard_count == 0)
+        THROW("Shard (Batch) size cannot be set to 0")
+    return (_file_id / (_batch_count)) % _shard_count;
+}
+
+Reader::Status TFRecordReader::read_image_names(std::ifstream &file_contents, uint file_size)
+{
+    auto ret = Reader::Status::OK;
+    uint length;
+    size_t uint64_size, uint32_size;
+    uint64_t data_length;
+    uint32_t length_crc, data_crc;
+    uint64_size = sizeof(uint64_t);
+    uint32_size = sizeof(uint32_t);
+    while (!_last_rec)
+    {
+        length = file_contents.tellg();
+        char *header_length = new char[uint64_size];
+        char *header_crc = new char[uint32_size];
+        char *footer_crc = new char[uint32_size];
+        file_contents.read(header_length, uint64_size);
+        if(!file_contents)
+            THROW("TFRecordReader: Error in reading TF records")
+        file_contents.read(header_crc, uint32_size);
```

Comment:
Codacy found an issue: [Check buffer boundaries if used in a loop including recursive loops (CWE-120, CWE-20).  ](https://app.codacy.com/gh/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/pullRequest?prid=5763058)

Reply:
This is fstream read. We are evaluating the read status in the next line.
if(!file_contents)
THROW("TFMetaDataReader: Error in reading TF records")
This can be ignored.